packages/flutter/test/widgets/navigator_replacement_test.dart

```diff
@@ -39,42 +42,205 @@ void main() {
     expect(find.text('home'), findsOneWidget);
   });
 
-  testWidgets('pushAndRemoveUntil', (WidgetTester tester) async {
-    await tester.pumpWidget(MaterialApp(
-      home: const Material(child: Text('home')),
-      routes: <String, WidgetBuilder>{
-        '/a': (BuildContext context) => const Material(child: Text('a')),
-        '/b': (BuildContext context) => const Material(child: Text('b')),
-      },
-    ));
-
-    final NavigatorState navigator = tester.state(find.byType(Navigator));
-    navigator.pushNamed('/a');
-    await tester.pumpAndSettle();
-
-    expect(find.text('home', skipOffstage: false), findsOneWidget);
-    expect(find.text('a', skipOffstage: false), findsOneWidget);
-    expect(find.text('b', skipOffstage: false), findsNothing);
-
-    navigator.pushNamedAndRemoveUntil('/b', (Route<dynamic> route) => false);
-    await tester.pumpAndSettle();
-
-    expect(find.text('home', skipOffstage: false), findsNothing);
-    expect(find.text('a', skipOffstage: false), findsNothing);
-    expect(find.text('b', skipOffstage: false), findsOneWidget);
-
-    navigator.pushNamed('/');
-    await tester.pumpAndSettle();
-
-    expect(find.text('home', skipOffstage: false), findsOneWidget);
-    expect(find.text('a', skipOffstage: false), findsNothing);
-    expect(find.text('b', skipOffstage: false), findsOneWidget);
-
-    navigator.pushNamedAndRemoveUntil('/a', ModalRoute.withName('/b'));
-    await tester.pumpAndSettle();
-
-    expect(find.text('home', skipOffstage: false), findsNothing);
-    expect(find.text('a', skipOffstage: false), findsOneWidget);
-    expect(find.text('b', skipOffstage: false), findsOneWidget);
+  group('pushAndRemoveUntil', () {
+
+    testWidgets('notifies appropriately', (WidgetTester tester) async {
+      final TestObserver observer = TestObserver();
+      final Widget myApp = MaterialApp(
+        home: const Material(child: Text('home')),
+        routes: <String, WidgetBuilder>{
+          '/a': (BuildContext context) => const Material(child: Text('a')),
+          '/b': (BuildContext context) => const Material(child: Text('b')),
+        },
+        navigatorObservers: <NavigatorObserver>[observer],
+      );
+
+      await tester.pumpWidget(myApp);
+
+      final NavigatorState navigator = tester.state(find.byType(Navigator));
+      navigator.pushNamed('/a');
+      await tester.pumpAndSettle();
+
+      expect(find.text('home', skipOffstage: false), findsOneWidget);
+      expect(find.text('a', skipOffstage: false), findsOneWidget);
+      expect(find.text('b', skipOffstage: false), findsNothing);
+
+      final List<String> log = <String>[];
+      observer
+        ..onPushed = (Route<dynamic> route, Route<dynamic> previousRoute) {
+          expectSync(route is PageRoute && route.settings.name == '/b', isTrue);
```

Comment:
Instead of having the expects in here you could just call `log.add('${route.settings.name} pushed, previous route: ${previousRoute.settings.name}')` and then verify everything at the end.

Reply:
Same for the onRemoved

Reply:
Pretty slick!

Reply:
Pretty slick! :)